runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkDefaultExecutableStageContext.java

```diff
@@ -64,19 +71,60 @@ public void close() throws Exception {
     jobBundleFactory.close();
   }
 
-  enum ReferenceCountingFactory implements Factory {
-    REFERENCE_COUNTING;
+  private static class JobFactoryState {
+    private final AtomicInteger counter = new AtomicInteger(0);
+    private final List<ReferenceCountingFlinkExecutableStageContextFactory> factories =
+        new ArrayList<>();
+    private final int maxFactories;
 
-    private static final ReferenceCountingFlinkExecutableStageContextFactory actualFactory =
-        ReferenceCountingFlinkExecutableStageContextFactory.create(
-            FlinkDefaultExecutableStageContext::create);
+    private JobFactoryState(int maxFactories) {
+      if (maxFactories == 0) {
+        // Default to num_cores - 1 so that we leave some resources available for the java process
+        this.maxFactories = Math.max(Runtime.getRuntime().availableProcessors() - 1, 1);
+      } else {
+        this.maxFactories = maxFactories;
+      }
+    }
+
+    private synchronized FlinkExecutableStageContext.Factory getFactory() {
+      int count = counter.getAndIncrement();
+
+      if (count < maxFactories) {
+        factories.add(
+            ReferenceCountingFlinkExecutableStageContextFactory.create(
+                FlinkDefaultExecutableStageContext::create));
+      }
+
+      return factories.get(count % maxFactories);
+    }
+  }
+
+  enum MultiInstanceFactory implements Factory {
+    MULTI_INSTANCE;
+
+    // This map should only ever have a single element, as each job will have its own
+    // classloader and therefore its own instance of MultiInstanceFactory.INSTANCE. This
+    // code supports multiple JobInfos in order to provide a sensible implementation of
+    // Factory.get(JobInfo), which in theory could be called with different JobInfos.
+    private static final ConcurrentMap<String, JobFactoryState> jobFactories =
+        new ConcurrentHashMap<>();
 
     @Override
     public FlinkExecutableStageContext get(JobInfo jobInfo) {
-      return actualFactory.get(jobInfo);
+      JobFactoryState state =
+          jobFactories.computeIfAbsent(
+              jobInfo.jobId(),
+              k -> {
+                PortablePipelineOptions portableOptions =
+                    PipelineOptionsTranslation.fromProto(jobInfo.pipelineOptions())
+                        .as(PortablePipelineOptions.class);
+
+                return new JobFactoryState(
+                    MoreObjects.firstNonNull(portableOptions.getSdkWorkerParallelism(), 1L)
```

Comment:
If we make the above change from `Long` to `long` with default `-1`, this would become:
```suggestion
                    Math.max(portableOptions.getSdkWorkerParallelism(), 1L)
```